### ISSUE TITLE
is there priority over data within server_group?

### ISSUE THREAD

**Issue by lf777** [jacksontj/promxy#238]
hi, my HA setup:
```
                                               /  prometheus (p1) <== \ 
                                             /                                               \
virtual machine (vm)<==                                                   promxy (pxy) <== grafana (graf)
                                             \                                               /
                                               \  prometheus (p2) <== /

```

sequence of events for flow A:
- p2 goes down, graf continues to observe vm as reported by p1.
- p2 recovers, graph on graf continues.

result is as expected for flow A.

sequence of events for flow B:
- p1 goes down, graf continues to observe vm as reported by p2.
- p1 recovers, graph on graf changes for the time period when p1 was down. for the period where p1 was down,it is as if the readings were extrapolated (from p1) to fill up the missing gap; and the readings originally from p2 were replaced by the extrapolated values.

result is not as expected for flow B.

is there any priority for promxy when choose data over p1 or p2?

promxy config is as such:
```
global:
  evaluation_interval: 5s

rule_files:
- "*rule"

promxy:
  server_groups:
    - static_configs:
        - targets:
          - p1:9090
          - p2:9090
      labels:
        sg: promxy_p
      anti_affinity: 10s
      http_client:
        tls_config:
          insecure_skip_verify: true
      ignore_error: true
```

thanks for any advice.

**Comment by jacksontj** [comment]
Promxy's merge logic today is [here](https://github.com/jacksontj/promxy/blob/master/pkg/promclient/multi_api.go#L547). What this translates to is that promxy will send requests to all of the downstreams in the severgroup and merge the results back. The order in which the merge happens is completely based on the order that the results come back in, since the queries are all sent roughly parallel it should be more-or-less random.

So in your "flow B" case if p1 returns first then p1's data would be "filled" with p2's data,  whereas if p2 returned first p1's data would be merged in (given that it doesn't have a hole, the data would in reality be ignored.).

The only way that I can see an "interpolation" happening is if the outage was smaller than the `LookbackDelta, in this case when promxy sends a `query_range` to p1 it might return points for that "gap" -- if so promxy would treat that data returned as valid.

Hopefully that answers the behavior question, if its still unclear or you have more questions feel free to ask :)

**Comment by lf777** [comment]
Hi, tested again with v0.0.55 to specify lookback delta from command line, such that the outage duration is greater than the delta.

with
```
promxy:
  server_groups:
    - static_configs:
        - targets:
          - p1:9090
          - p2:9090
```
pxy's picture on graf is similar to p1's.


and with
```
promxy:
  server_groups:
    - static_configs:
        - targets:
          - p2:9090
          - p1:9090
```
pxy's picture on graf is similar to p2's.

strange, sequence of entry in config file seems to affect the results.

**Comment by jacksontj** [comment]
@lf777 if your 2 downstream targets are very fast it could make sense. Although the queries are "in parallel" it does launch the goroutines in serial -- so assuming the response is very fast the order in the list does give it an edge in the race condition.

> pxy's picture on graf is similar to p1's.

Meaning it still has the `extrapolated` looking points? If so, I'd be curious to see a pcap + trace logs of the query (so we can see what data is getting returned to promxy and what it returns).

**Comment by lf777** [comment]
Yes, the downstream targets are fast.
pxy and p1 are currently colocated on the same machine.
Would it still help to look at the logs?

**Comment by jacksontj** [comment]
At this point it seems that this is behaving as we'd expect given the current code ("base" is whoever returns first, and we merge from there). It seems though that we'd prefer (or optionally prefer) to use the "largest" response as the base and merge from there (or some other mechanism to start reduce the "fill" required).

This would require at a minimum changes in the [merge function](https://github.com/jacksontj/promxy/blob/master/pkg/promhttputil/merge.go#L74) and might require some logic changes in the [multi_api client](https://github.com/jacksontj/promxy/blob/master/pkg/promclient/multi_api.go) as well.


Short version: basically your ask is to add priority for data in a servergroup, where priority is defined by "has the most data", and that should be doable.

I'll sit on this for a couple days to mull it over, but it sounds reasonable.
